src/library/scala/collection/immutable/Vector.scala

```diff
@@ -758,13 +829,135 @@ final class VectorBuilder[A]() extends ReusableBuilder[A, Vector[A]] with Vector
     this
   }
 
+  override def addAll(xs: IterableOnce[A]): this.type = {
+
+    xs match {
+      case v: Vector[A] if this.isEmpty && v.length >= 32
+        => // Before merging this line should be moved to the line above, only like this to allow quick commenting out.
+//        && false => // this is only here to be able to quickly disable the new functionality by uncommenting. Remove this before merging!!!
+      if (v.nonEmpty) {
+        depth = v.depth
+        blockIndex = (v.endIndex - 1) & ~31
+        lo = v.endIndex - blockIndex
+        startIndex = v.startIndex
+
+        /** A modified version of nullSlotAndCopy` which writes to a destination array rather than create a new copy */
+        def nullSlotAndOverwrite[T <: AnyRef](array: Array[Array[T]], index: Int, destination: Array[AnyRef]): destination.type = {
+          val x = array(index)
+          array(index) = null
+          x.copyToArray(destination, 0)
+          destination
+        }
+
+        /** A modified version of gotoPositionWritable1, which avoids creating a new array for display0, by reusing the passed array as a copy destination */
+        def goToPositionWritableReuseDisplay0(initialDisplay0: Array[AnyRef]): Unit = {
```

Comment:
Do we need separate versions of `gotoPositionWritable1` and `nullSlotAndCopy`? Is there a performance penalty when modifying the original versions to pass in a possibly null array?

Reply:
That is a good idea!

Reply:
I have now implemented these two methods in the suggested way, with a possibly null array argument to the existing methods.